File: src/main/java/investLee/platform/ecommerce/domain/entity/CategoryEntity.java
package investLee.platform.ecommerce.domain.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "categories")
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class CategoryEntity {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String slug;
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private CategoryEntity parent;

    private int level;

    private String imageUrl;

    @OneToMany(mappedBy = "category")
    @Builder.Default
    private List<ProductCategoryEntity> productCategories = new ArrayList<>();
}

File: src/modules/categories/entities/category.entity.ts
import { z } from 'zod';

// 기본 카테고리 스키마
export const CategorySchema = z.object({
  id: z.number(),
  name: z.string(),
  slug: z.string(),
  description: z.string().nullable(),
  parentId: z.number().nullable(),
  level: z.number(),
  imageUrl: z.string().nullable(),
});

// 관계를 포함한 카테고리 스키마
export const CategoryWithRelationsSchema = CategorySchema.extend({
  parent: z.lazy(() => CategorySchema).optional(),
  children: z.array(z.lazy(() => CategoryWithRelationsSchema)).optional(),
});

// 타입 추론
export type Category = z.infer<typeof CategorySchema>;
export type CategoryWithRelations = z.infer<typeof CategoryWithRelationsSchema>;
